feat(new transform): new `compound` transform

prognant opened the pull request:
closes DataDog/vector-enterprise#156

New transform to aggregate multiple transforms into a single one:

```
[transforms.chain]
type = "compound"
inputs = [ "xyz" ]

[transforms.pipelines.nested.step_1]
type = "remap"
source = '''
.foo = "bar" 
'''

[transforms.pipelines.nested.step_2]
type = "remap"
source = '''
.foo, _ =  .foo + "baz"
'''

```

Any tranforms in the chain can be reference as an input using `<transform_id>.<subtransform_id>` (e.g. `chain.step_2` to pick up events after being processed by step_1 & step_2

binarylogic commented:
@prognant which issue does this close?

binarylogic commented:
Mind nothing that in the PR description then so it closes when this is merged? It helps with referencing.

binarylogic commented:
@prognant small nit, can you avoid force pushing unless necessary? It helps when following along with changes. Thanks.

spencergilbert commented:
Looks like `docs/cue/reference/.urls.cue.swp` was accidentally added

jszwedko commented on src/config/unit_test.rs:
```diff
@@ -1447,4 +1447,51 @@ mod tests {
     }
+
+    #[tokio::test]
+    async fn test_expand() {
```
I think this test would better fit into `tests/behavior/transform/`. I think the tests in this file are more about testing unit test configuration than validating transforms.

prognant replied:
I never looked at those tests. This make much more sense to put it there. Will relocate.

jszwedko commented on src/transforms/compound.rs:
```diff
@@ -0,0 +1,93 @@
+        if !self.nested.is_empty() {
+            Ok(Some((self.nested.clone(), ExpandType::Serial)))
+        } else {
+            Err("must specify at least one transform".into())
```
Is this a strict requirement because of the implementation?

I realize it isn't functionally useful, but I could see allowing users to have an empty `compound` transform that just passes through events. This could be useful if, for example, they wanted to temporarily comment out all of the serial transforms.

prognant replied:
Bouncing on the other comment, I think this could be addressed by having a top level passthrough transform. Ideally this noop transform would bear the name of the `compound` transform as it is defined in the config, it would also be usable to generate top level metrics as you suggested above, WDYT ?

jszwedko replied:
👍 yeah, a noop transform seems like it would help for that case. We can follow up on it later though; I think it'll be rare that a user would hit this.

jszwedko commented on docs/cue/reference/components/transforms/compound.cue:
```diff
@@ -0,0 +1,113 @@
+			output: [
+				{
+					log: {
+						message: "[info] I'm a normal info log"
```
Shouldn't this be the other way around? That the debug log selected?

prognant replied:
Nice catch !

jszwedko reviewed (commented):
Nice work on this!

I tested it out locally and it seems to work well.

I verified the metrics are correctly generated for subtransforms:

```
vector_events_out_total{component_kind="transform",component_name="foo.step2",component_type="add_fields"} 79863 1627502953819
vector_events_out_total{component_kind="transform",component_name="foo.step1",component_type="add_fields"} 79887 1627502953819
```

I am wondering if we could still generate top-level metrics for the whole transform though. That is, in addition to the per nested transform metrics shown above, we would also have:

```
vector_events_out_total{component_kind="transform",component_name="foo",component_type="compound"} 79887 1627502953819
```

?

I'm unsure how invasive a change that would be though given the way this transform is leveraging the existing expand transform functionality (which makes sense!).

I'd like to get @lukesteensen's 👀 on this too.

lukesteensen commented on src/transforms/compound.rs:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Any
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Any
```
I was going to say that these two probably need to return the types from the first and last transforms in `nested`, but I think we're actually ok because expansion happens before type checking. Still might be worth having a test that doing things like `log -> metric` return the same error that it would outside of the compound transform.

lukesteensen reviewed (approved):
This is a really clever way to implement this! I like that it's very lightweight and just reuses existing Vector functionality.

I think there's a chance that we may eventually want something that actually runs all the transforms inline with some additional logic (e.g. in/out metrics for the whole set, inline filtering, etc), either for feature reasons or maybe performance reasons (I'd be curious to benchmark the concurrency of multiple components vs the simplicity of cutting out channels), but this strikes me as a perfectly good initial implementation.

prognant commented:
I investigated the possibility to have a top level transform to get metrics tags with the top level name seems overkill, I added a note in the doc that explicitly says that sub-transforms generates metrics as if they were configured independently.
I added extra tests, including type mismatch.
